shotover-proxy/src/codec/cassandra.rs

```diff
@@ -278,54 +544,136 @@ impl Encoder<Messages> for CassandraEncoder {
+        let handshake_complete = self.handshake_complete.load(Ordering::Relaxed);
+
         for m in item {
             let start = dst.len();
-            let compression = m.codec_state.as_cassandra();
-
-            // TODO: always check if cassandra message
-            match m.into_encodable(MessageType::Cassandra)? {
-                Encodable::Bytes(bytes) => {
-                    // check if the message is a startup message and set the codec's compression
-                    {
-                        let opcode = Opcode::try_from(bytes[4])?;
-                        if Opcode::Startup == opcode {
-                            if let CassandraFrame {
-                                operation: CassandraOperation::Startup(startup),
-                                ..
-                            } = CassandraFrame::from_bytes(bytes.clone(), Compression::None)?
-                            {
-                                set_compression(&mut self.compression, &startup);
-                            };
-                        }
-                    }
+            self.encode_frame(dst, m, version, handshake_complete)?;
+            tracing::debug!(
+                "{}: outgoing cassandra message:\n{}",
+                self.direction,
+                pretty_hex::pretty_hex(&&dst[start..])
+            );
+        }
+        Ok(())
+    }
+}
 
-                    dst.extend_from_slice(&bytes)
+impl CassandraEncoder {
+    fn encode_frame(
+        &mut self,
+        dst: &mut BytesMut,
+        m: Message,
+        version: Version,
+        handshake_complete: bool,
+    ) -> Result<()> {
+        match (version, handshake_complete) {
+            (Version::V5, true) => {
+                // write envelope header with dummy values for those we cant calculate till after we write the message
+                let header_start = dst.len();
+
+                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
+                let payload_start = dst.len();
+
+                self.encode_envelope(dst, m)?;
+
+                //measure length of message and calculate crc24 and overwrite frame header values
+                let mut payload_len = (dst.len() - payload_start) as u64;
+
+                if true {
+                    // TODO if self_contained
+                    payload_len |= 1 << 17;
                 }
-                Encodable::Frame(frame) => {
-                    // check if the message is a startup message and set the codec's compression
-                    if let Frame::Cassandra(CassandraFrame {
-                        operation: CassandraOperation::Startup(startup),
-                        ..
-                    }) = &frame
-                    {
-                        set_compression(&mut self.compression, startup);
-                    };
 
-                    let buffer = frame.into_cassandra().unwrap().encode(compression);
+                put3b(&mut dst[header_start..], payload_len as i32);
```

Comment:
I was curious so I checked the assembly of both and the copy_from_slice approach seems to produce better assembly, I assume because it doesnt have to do a bounds check on every index into buffer in put3b
![image](https://user-images.githubusercontent.com/5120858/228157072-7366580a-e65c-4a21-8a9b-f87a8eef1712.png)